src/github.com/matrix-org/dendrite/roomserver/types/types.go

```diff
@@ -119,3 +125,19 @@ type StateEntryList struct {
 	StateBlockNID StateBlockNID
 	StateEntries  []StateEntry
 }
+
+// A RoomRecentEventsUpdater is used to update the recent events in a room.
+// (On postgresql this wraps a database transaction that holds a "FOR UPDATE"
+//  lock on the a row holding the latest events for the room.)
+type RoomRecentEventsUpdater interface {
+	// Store the previous events referenced by an event.
+	StorePreviousEvents(eventNID EventNID, previousEventReferences []gomatrixserverlib.EventReference) error
+	// Check whether the eventReference is already referenced by another matrix event.
+	IsReferenced(eventReference gomatrixserverlib.EventReference) (bool, error)
+	// Set the list of latest events for the room.
+	SetLatestEvents(roomNID RoomNID, latest []StateAtEventAndReference) error
```

Comment:
Naming: Is there some distinction between `StoreXXX` and `SetXXX` that you're making here? "Store" looks like it is INSERTing, and `Set` looks like it is updating, but it's completely unclear. If I try to set the same thing twice will they append? Clobber? Error out?

I don't mind exposing a commit/rollback interface like this, provided it is clear what actions can be taken, and what a compliant implementation of `RoomRecentEventsUpdater` should do.